research/deep_speech/deep_speech.py

```diff
@@ -0,0 +1,269 @@
+
+  # Number of label classes. Label string is "abcdefghijklmnopqrstuvwxyz' -"
+  num_classes = len(train_speech_dataset.speech_labels)
+
+  # Input shape of each data example: [time_steps (T), feature_bins(F), channel(C)]
+  # Channel is set as 1 by default.
+  input_shape = (None, train_speech_dataset.num_feature_bins, 1)
+
+  # Create deep speech model and convert it to Estimator
+  tf.logging.info("Creating Estimator from Keras model...")
+  keras_model = deep_speech_model.DeepSpeech(
+      input_shape, flags_obj.rnn_hidden_layers, flags_obj.rnn_type,
+      flags_obj.is_bidirectional, flags_obj.rnn_hidden_size,
+      flags_obj.rnn_activation, num_classes, flags_obj.use_bias)
+  # Check model structure
+  tf.logging.info(keras_model.summary(line_length=100))
+
+  # Convert to estimator
+  num_gpus = flags_core.get_num_gpus(flags_obj)
+  estimator = convert_keras_to_estimator(keras_model, num_gpus)
+
+  # Benchmark logging
+  run_params = {
+      "batch_size": flags_obj.batch_size,
+      "train_epochs": flags_obj.train_epochs,
+      "rnn_hidden_size": flags_obj.rnn_hidden_size,
+      "rnn_hidden_layers": flags_obj.rnn_hidden_layers,
+      "rnn_activation": flags_obj.rnn_activation,
+      "rnn_type": flags_obj.rnn_type,
+      "is_bidirectional":flags_obj.is_bidirectional,
+      "use_bias": flags_obj.use_bias
+  }
+
+  dataset_name = "LibriSpeech"
+  benchmark_logger = logger.get_benchmark_logger()
+  benchmark_logger.log_run_info("deep_speech", dataset_name, run_params,
+                                test_id=flags_obj.benchmark_test_id)
+
+  train_hooks = hooks_helper.get_train_hooks(
+      flags_obj.hooks,
+      batch_size=flags_obj.batch_size)
+
+  per_device_batch_size = distribution_utils.per_device_batch_size(
+      flags_obj.batch_size, num_gpus)
+
+  def input_fn_train():
+    return dataset.input_fn(
+        True, per_device_batch_size, train_speech_dataset)
+
+  def input_fn_eval():
+    return dataset.input_fn(
+        False, per_device_batch_size, eval_speech_dataset)
+
+  total_training_cycle = (flags_obj.train_epochs //
+                          flags_obj.epochs_between_evals)
+  for cycle_index in range(total_training_cycle):
+    tf.logging.info("Starting a training cycle: %d/%d",
+                    cycle_index + 1, total_training_cycle)
+
+    estimator.train(input_fn=input_fn_train, hooks=train_hooks)
+
+    # Evaluate (TODO)
+    # tf.logging.info("Starting to evaluate.")
+
+    # eval_results = evaluate_model(
+    #     estimator, keras_model, data_set.speech_labels, [], input_fn_eval)
+
+    # benchmark_logger.log_evaluation_result(eval_results)
+
+    # if model_helpers.past_stop_threshold(
+    #   flags_obj.stop_threshold, eval_results["accuracy"]):
+    #   break
+
+  if flags_obj.export_dir is not None:
+    # Exports a saved model for the given classifier.
+    input_receiver_fn = export.build_tensor_serving_input_receiver_fn(
+        shape, batch_size=flags_obj.batch_size)
```

Comment:
I suspect this will raise an error, as shape is undefined. What is the expected input shape/format when predicting?